File: flutter/llx_flutter/android/src/main/kotlin/com/inferx/llx_flutter/LlxFlutterPlugin.kt
package com.inferx.llx_flutter

import com.inferx.llx.LLX
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executors

/** LlxFlutterPlugin */
class LlxFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private val mainHandler = Handler(Looper.getMainLooper())
    // 串行执行，避免底层非线程安全；如需并发，可改为线程池
    private val executor = Executors.newSingleThreadExecutor()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "llx_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "initBackend" -> {
                    executor.execute {
                        try {
                            LLX.nativeInitBackend()
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "freeBackend" -> {
                    executor.execute {
                        try {
                            LLX.nativeFreeBackend()
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "modelLoad" -> {
                    val path = call.argument<String>("path")
                        ?: return result.error("INVALID_ARGUMENT", "path is required", null)
                    executor.execute {
                        try {
                            val handle = LLX.nativeModelLoad(path)
                            mainHandler.post { result.success(handle) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "modelFree" -> {
                    val handle = call.argument<Number>("modelHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "modelHandle is required", null)
                    executor.execute {
                        try {
                            LLX.nativeModelFree(handle)
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionCreate" -> {
                    val modelHandle = call.argument<Number>("modelHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "modelHandle is required", null)
                    val nCtx = call.argument<Number>("nCtx")?.toInt() ?: 8192
                    val nThreads = call.argument<Number>("nThreads")?.toInt() ?: 0
                    executor.execute {
                        try {
                            val handle = LLX.nativeSessionCreate(modelHandle, nCtx, nThreads)
                            mainHandler.post { result.success(handle) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionFree" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    executor.execute {
                        try {
                            LLX.nativeSessionFree(handle)
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionInitFromText" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val text = call.argument<String>("text")
                        ?: return result.error("INVALID_ARGUMENT", "text is required", null)
                    val formatChat = call.argument<Boolean>("formatChat") ?: true
                    val nLen = call.argument<Number>("nLen")?.toInt() ?: 1024
                    executor.execute {
                        try {
                            val tokCount = LLX.nativeSessionInitFromText(handle, text, formatChat, nLen)
                            mainHandler.post { result.success(tokCount) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionInitFromMessagesJson" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val messagesJson = call.argument<String>("messagesJson")
                        ?: return result.error("INVALID_ARGUMENT", "messagesJson is required", null)
                    val nLen = call.argument<Number>("nLen")?.toInt() ?: 1024
                    executor.execute {
                        try {
                            val tokCount = LLX.nativeSessionInitFromMessagesJson(handle, messagesJson, nLen)
                            mainHandler.post { result.success(tokCount) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionKvClear" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    executor.execute {
                        try {
                            LLX.nativeSessionKvClear(handle)
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionStep" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val nLen = call.argument<Number>("nLen")?.toInt() ?: 1024
                    executor.execute {
                        try {
                            val stepResult = LLX.nativeSessionStep(handle, nLen)
                            val resultMap = mapOf(
                                "text" to stepResult.text,
                                "finished" to stepResult.finished
                            )
                            mainHandler.post { result.success(resultMap) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "chatCompleteJson" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val requestJson = call.argument<String>("requestJson")
                        ?: return result.error("INVALID_ARGUMENT", "requestJson is required", null)
                    executor.execute {
                        try {
                            val responseJson = LLX.nativeChatCompleteJson(handle, requestJson)
                            mainHandler.post { result.success(responseJson) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "systemInfo" -> {
                    executor.execute {
                        try {
                            val info = LLX.nativeSystemInfo()
                            mainHandler.post { result.success(info) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "bench" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val pp = call.argument<Number>("pp")?.toInt() ?: 8
                    val tg = call.argument<Number>("tg")?.toInt() ?: 4
                    val pl = call.argument<Number>("pl")?.toInt() ?: 1
                    val nr = call.argument<Number>("nr")?.toInt() ?: 1
                    executor.execute {
                        try {
                            val benchResult = LLX.nativeBench(handle, pp, tg, pl, nr)
                            mainHandler.post { result.success(benchResult) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionLoadLora" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val loraPath = call.argument<String>("loraPath")
                        ?: return result.error("INVALID_ARGUMENT", "loraPath is required", null)
                    val scale = call.argument<Number>("scale")?.toFloat() ?: 1.0f
                    executor.execute {
                        try {
                            val success = LLX.nativeSessionLoadLora(handle, loraPath, scale)
                            mainHandler.post { result.success(success) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionAddLora" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val loraPath = call.argument<String>("loraPath")
                        ?: return result.error("INVALID_ARGUMENT", "loraPath is required", null)
                    val scale = call.argument<Number>("scale")?.toFloat() ?: 1.0f
                    executor.execute {
                        try {
                            val success = LLX.nativeSessionAddLora(handle, loraPath, scale)
                            mainHandler.post { result.success(success) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionUpdateLoraScale" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val loraPath = call.argument<String>("loraPath")
                        ?: return result.error("INVALID_ARGUMENT", "loraPath is required", null)
                    val scale = call.argument<Number>("scale")?.toFloat() ?: 1.0f
                    executor.execute {
                        try {
                            val success = LLX.nativeSessionUpdateLoraScale(handle, loraPath, scale)
                            mainHandler.post { result.success(success) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionRemoveLora" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    val loraPath = call.argument<String>("loraPath")
                        ?: return result.error("INVALID_ARGUMENT", "loraPath is required", null)
                    executor.execute {
                        try {
                            LLX.nativeSessionRemoveLora(handle, loraPath)
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "sessionClearLora" -> {
                    val handle = call.argument<Number>("sessionHandle")?.toLong()
                        ?: return result.error("INVALID_ARGUMENT", "sessionHandle is required", null)
                    executor.execute {
                        try {
                            LLX.nativeSessionClearLora(handle)
                            mainHandler.post { result.success(null) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                "lastError" -> {
                    executor.execute {
                        try {
                            val error = LLX.nativeLastError()
                            mainHandler.post { result.success(error) }
                        } catch (e: Exception) {
                            mainHandler.post { result.error("ERROR", e.message ?: "Unknown error", null) }
                        }
                    }
                }
                else -> {
                    result.notImplemented()
                }
            }
        } catch (e: Exception) {
            result.error("ERROR", e.message ?: "Unknown error", null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: flutter/llx_flutter/ios/Classes/LlxFlutterPlugin.swift
import Flutter
import UIKit

// Import Swift kit module defined in ../../../ios/llx-ios/Sources
// 同一 Pod 目标内直接访问公开类型

public class LlxFlutterPlugin: NSObject, FlutterPlugin {
  private static var channelName = "llx_flutter"

  // Handle maps
  private var models: [Int64: InferxModel] = [:]
  private var sessions: [Int64: InferxSession] = [:]
  private var nextHandle: Int64 = 1
  private let queue = DispatchQueue(label: "com.inferx.llx_flutter.serial")

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
    let instance = LlxFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initBackend":
      queue.async {
        InferxBackend.initBackend()
        DispatchQueue.main.async { result(nil) }
      }

    case "freeBackend":
      queue.async {
        InferxBackend.freeBackend()
        DispatchQueue.main.async { result(nil) }
      }

    case "modelLoad":
      guard let args = call.arguments as? [String: Any], let path = args["path"] as? String else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "path is required", details: nil))
      }
      queue.async {
        do {
          let m = try InferxModel(path: path)
          let h = self.allocateHandle()
          self.models[h] = m
          DispatchQueue.main.async { result(h) }
        } catch {
          DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) }
        }
      }

    case "modelFree":
      guard let args = call.arguments as? [String: Any], let mh = (args["modelHandle"] as? NSNumber)?.int64Value else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "modelHandle is required", details: nil))
      }
      queue.async {
        self.models.removeValue(forKey: mh)
        DispatchQueue.main.async { result(nil) }
      }

    case "sessionCreate":
      guard let args = call.arguments as? [String: Any],
            let mh = (args["modelHandle"] as? NSNumber)?.int64Value,
            let model = models[mh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let nCtx: Int32 = (args["nCtx"] as? NSNumber)?.int32Value ?? 8192
      let nThreads: Int32 = (args["nThreads"] as? NSNumber)?.int32Value ?? 0
      queue.async {
        do {
          let sess = try model.createSession(params: SessionParams(nCtx: nCtx, nThreads: nThreads))
          let sh = self.allocateHandle()
          self.sessions[sh] = sess
          DispatchQueue.main.async { result(sh) }
        } catch {
          DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) }
        }
      }

    case "sessionFree":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "sessionHandle is required", details: nil))
      }
      queue.async {
        self.sessions.removeValue(forKey: sh)
        DispatchQueue.main.async { result(nil) }
      }

    case "sessionInitFromText":
      guard let args = call.arguments as? [String: Any],
            let sh = (args["sessionHandle"] as? NSNumber)?.int64Value,
            let text = args["text"] as? String,
            let sess = sessions[sh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let formatChat = (args["formatChat"] as? NSNumber)?.boolValue ?? true
      let nLen = (args["nLen"] as? NSNumber)?.int32Value ?? 1024
      queue.async {
        do {
          let count = try sess.initFromText(text, formatChat: formatChat, maxLen: nLen)
          DispatchQueue.main.async { result(count) }
        } catch {
          DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) }
        }
      }

    case "sessionInitFromMessagesJson":
      guard let args = call.arguments as? [String: Any],
            let sh = (args["sessionHandle"] as? NSNumber)?.int64Value,
            let messagesJson = args["messagesJson"] as? String,
            let sess = sessions[sh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let nLen = (args["nLen"] as? NSNumber)?.int32Value ?? 1024
      queue.async {
        do {
          let count = try sess.initFromMessagesJSON(messagesJson, maxLen: nLen)
          DispatchQueue.main.async { result(count) }
        } catch {
          DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) }
        }
      }

    case "sessionKvClear":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "sessionHandle is required", details: nil))
      }
      queue.async {
        sess.clearKV()
        DispatchQueue.main.async { result(nil) }
      }

    case "sessionStep":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "sessionHandle is required", details: nil))
      }
      let nLen = (args["nLen"] as? NSNumber)?.int32Value ?? 1024
      queue.async {
        do {
          let (text, finished) = try sess.step(maxLen: nLen)
          DispatchQueue.main.async { result(["text": text, "finished": finished]) }
        } catch {
          DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) }
        }
      }

    case "chatCompleteJson":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh], let requestJson = args["requestJson"] as? String else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      queue.async {
        do { let s = try sess.chatComplete(requestJSON: requestJson); DispatchQueue.main.async { result(s) } }
        catch { DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) } }
      }

    case "systemInfo":
      queue.async { let s = InferxBackend.systemInfo(); DispatchQueue.main.async { result(s) } }

    case "bench":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let pp = (args["pp"] as? NSNumber)?.int32Value ?? 8
      let tg = (args["tg"] as? NSNumber)?.int32Value ?? 4
      let pl = (args["pl"] as? NSNumber)?.int32Value ?? 1
      let nr = (args["nr"] as? NSNumber)?.int32Value ?? 1
      queue.async {
        do { let s = try sess.bench(pp: pp, tg: tg, pl: pl, nr: nr); DispatchQueue.main.async { result(s) } }
        catch { DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) } }
      }

    case "sessionLoadLora":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh], let loraPath = args["loraPath"] as? String else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let scale = (args["scale"] as? NSNumber)?.floatValue ?? 1.0
      queue.async {
        do { let ok = try sess.loadLora(path: loraPath, scale: scale); DispatchQueue.main.async { result(ok) } }
        catch { DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) } }
      }

    case "sessionAddLora":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh], let loraPath = args["loraPath"] as? String else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let scale = (args["scale"] as? NSNumber)?.floatValue ?? 1.0
      queue.async {
        do { let ok = try sess.addLora(path: loraPath, scale: scale); DispatchQueue.main.async { result(ok) } }
        catch { DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) } }
      }

    case "sessionUpdateLoraScale":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh], let loraPath = args["loraPath"] as? String else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      let scale = (args["scale"] as? NSNumber)?.floatValue ?? 1.0
      queue.async {
        do { let ok = try sess.updateLoraScale(path: loraPath, scale: scale); DispatchQueue.main.async { result(ok) } }
        catch { DispatchQueue.main.async { result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil)) } }
      }

    case "sessionRemoveLora":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh], let loraPath = args["loraPath"] as? String else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "invalid args", details: nil))
      }
      queue.async {
        sess.removeLora(path: loraPath)
        DispatchQueue.main.async { result(nil) }
      }

    case "sessionClearLora":
      guard let args = call.arguments as? [String: Any], let sh = (args["sessionHandle"] as? NSNumber)?.int64Value, let sess = sessions[sh] else {
        return result(FlutterError(code: "INVALID_ARGUMENT", message: "sessionHandle is required", details: nil))
      }
      queue.async {
        sess.clearLora()
        DispatchQueue.main.async { result(nil) }
      }

    case "lastError":
      queue.async { let s = InferxBackend.lastError(); DispatchQueue.main.async { result(s) } }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func allocateHandle() -> Int64 { defer { nextHandle += 1 }; return nextHandle }
}


